cmd/fund/fund.go

```diff
@@ -0,0 +1,268 @@
+package fund
+
+import (
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	_ "embed"
+
+	"github.com/chenzhijie/go-web3"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/cobra"
+)
+
+var (
+	//go:embed usage.md
+	usage                  string
+	walletCount            int
+	fundingWalletPK        string
+	fundingWalletPublicKey *ecdsa.PublicKey
+	chainID                int
+	chainRPC               string
+	concurrencyLevel       int
+	walletFundingAmt       float64
+	walletFundingGas       uint64
+	verbosityEnabled       bool
+	nonceMutex             sync.Mutex
+	globalNonce            uint64
+	nonceInitialized       bool
+	outputFileFlag         string
+)
+
+// Wallet struct to hold public key, private key, and address
+type Wallet struct {
+	PublicKey  *ecdsa.PublicKey
+	PrivateKey *ecdsa.PrivateKey
+	Address    common.Address
+}
+
+func generateNonce(web3Client *web3.Web3) (uint64, error) {
+	nonceMutex.Lock()
+	defer nonceMutex.Unlock()
+
+	if nonceInitialized {
+		globalNonce++
+	} else {
+		// Derive the public key from the funding wallet's private key
+		fundingWalletECDSA, ecdsaErr := crypto.HexToECDSA(fundingWalletPK)
+		if ecdsaErr != nil {
+			log.Error().Err(ecdsaErr).Msg("Error getting ECDSA from funding wallet private key")
+			return 0, ecdsaErr
+		}
+
+		fundingWalletPublicKey = &fundingWalletECDSA.PublicKey
+		// Convert ecdsa.PublicKey to common.Address
+		fundingAddress := crypto.PubkeyToAddress(*fundingWalletPublicKey)
+
+		nonce, err := web3Client.Eth.GetNonce(fundingAddress, nil)
+		if err != nil {
+			log.Error().Err(err).Msg("Error getting nonce")
+			return 0, err
+		}
+		globalNonce = nonce
+		nonceInitialized = true
+	}
+
+	return globalNonce, nil
+}
+
+func generateWallets(numWallets int) ([]Wallet, error) {
+	wallets := make([]Wallet, 0, numWallets)
+
+	for i := 0; i < numWallets; i++ {
+		account, err := crypto.GenerateKey()
+		if err != nil {
+			log.Error().Err(err).Msg("Error generating key")
+			return nil, err
+		}
+
+		addr := crypto.PubkeyToAddress(account.PublicKey)
+		wallet := Wallet{
+			PublicKey:  &account.PublicKey,
+			PrivateKey: account,
+			Address:    addr,
+		}
+
+		wallets = append(wallets, wallet)
```

Comment:
```suggestion
		wallets[i] = wallet
```

Reply:
this throws:
`panic: runtime error: index out of range [0] with length 0`

will stick with current version for time being

Reply:
Oh! That's because the allocation line is not correct. It should be `wallets := make([]Wallet, numWallets)`.

Reply:
fixed